### ISSUE TITLE
missing programatic migration path from `POST_FLAT_FILE_LISTINGS_DATA` to `JSON_LISTINGS_FEED`

### ISSUE THREAD

**Issue by johnkw** [amzn/selling-partner-api-models#4318]
There is currently no sample workflow for editing a listing using the new non-deprecated options, now that the `POST_FLAT_FILE_LISTINGS_DATA` update is deprecated.

Previously, the workflow was extremely simple:

1. Go to https://sellercentral.amazon.com/listing/cards/add-product?style=standalone and type in your product type.
2. Download the template spreadsheet, fill it out, save it as a CSV.
3. Upload via /feeds/2021-06-30/feeds.

Now however with `/feeds/2021-06-30/feeds` being deprecated, step 1 is immediately a problem.  There's no JSON equivalent to download on SellerCentral.  And there's no programmatic option for translating the current easily-created CSV into the newly-required JSON.

There is a [Product Flat File Mappings](https://m.media-amazon.com/images/G/01/VSSC/FlatFile-To-JSON-Mapping.xlsx) linked from https://developer-docs.amazon.com/sp-api/docs/mapping-product-attributes however it's missing some fields.  For example it's missing all of the `other_image_url*` fields.

Of special note, there has actually been a third-party SaaS which has seized upon the opportunity caused by this SPAPI bug, and created a proprietary/undocumented system to translate the CSV into JSON: https://github.com/amzn/selling-partner-api-models/issues/572#issuecomment-2453912368

Prior related issues: https://github.com/amzn/selling-partner-api-samples/issues/56 https://github.com/amzn/selling-partner-api-models/issues/547 https://github.com/amzn/selling-partner-api-models/issues/572

**Comment by ritikacd** [comment]
Hi @johnkw 

Other than the recommendations provided on the issue #[547](https://github.com/amzn/selling-partner-api-models/issues/547) and #[56](https://github.com/amzn/selling-partner-api-samples/issues/56), we suggest reviewing the different [migration strategies](https://developer-docs.amazon.com/sp-api/docs/listings-migration-strategies) here that will you migrate your workflows from feed type POST_FLAT_FILE_LISTINGS_DATA to Listings Items APIs or JSON_LISTINGS_FEED feed type. In one of the migration strategies, you can see that you could retrieve existing listing created by old workflow using new Listings Items API. This strategy will help you understand the JSON schemas and how your mappings should be structured. 

You can also find the majority of the product and non product attributes mappings on [this](https://developer-docs.amazon.com/sp-api/docs/mapping-product-attributes) page. For `other_image_url*` fields, `other_product_image_locator_*` is the new field that can be used. 

Let us know if you still have issues. 

Thanks,
Ritika
Selling Partner Developer Services

**Comment by johnkw** [comment]
Why can't the Amazon documentation team just add the missing items from https://developer-docs.amazon.com/sp-api/docs/mapping-product-attributes into the https://m.media-amazon.com/images/G/01/VSSC/FlatFile-To-JSON-Mapping.xlsx file?

**Comment by ritikacd** [comment]
Hi @johnkw 

We’ve been investigating the missing mappings, as most of the mappings are present for legacy XML and flat feed types on this [page](https://developer-docs.amazon.com/sp-api/docs/mapping-product-attributes). Other than the image mappings, do you see any other mappings that are missing from a flatfile-to-json mapping file? 

Thanks,
Ritika
Selling Partner Developer Services

**Comment by johnkw** [comment]
`back_bottle_image_url
back_label_image_url
business_price
feed_product_type
front_bottle_image_url
fulfillment_center_id
fulfillment_latency
item_dimensions_unit_of_measure
main_image_url
map_price
net_content_volume1
net_content_volume2
net_content_volume_unit_of_measure1
net_content_volume_unit_of_measure2
offer_image1
offer_image2
offer_image3
offer_image4
offer_image5
offering_end_date
offering_start_date
other_image_url1
other_image_url2
other_image_url3
other_image_url4
other_image_url5
other_image_url6
other_image_url7
other_image_url8
package_contains_identifier
package_contains_quantity
package_dimensions_unit_of_measure
parent_sku
pricing_action
progressive_discount_lower_bound1
progressive_discount_lower_bound2
progressive_discount_lower_bound3
progressive_discount_type
progressive_discount_value1
progressive_discount_value2
progressive_discount_value3
quantity
quantity_lower_bound1
quantity_lower_bound2
quantity_lower_bound3
quantity_lower_bound4
quantity_lower_bound5
quantity_price1
quantity_price2
quantity_price3
quantity_price4
quantity_price5
quantity_price_type
relationship_type
restock_date
sale_end_date
sale_from_date
sale_price
shaft_style_type
standard_price
swatch_image_url
`

**Comment by johnkw** [comment]
Can we leave this open until this bug is fixed? The page you link to is not of any use here, but perhaps other people will see this bug and contribute additional information prior to the bug itself being fixed.

**Comment by ritikacd** [comment]
> Can we leave this open until this bug is fixed? The page you link to is not of any use here, but perhaps other people will see this bug and contribute additional information prior to the bug itself being fixed.

Hi @johnkw 
We have opened the issue to address your concern. We’d like to inform you that in the legacy feeds, some attributes were duplicated and have since been deprecated in the new Listings APIs. We are currently working on clarifying this information in the mapping file to ensure better understanding and ease of use. We appreciate your patience as we work to improve our documentation.

Thank you,
Ritika

**Comment by johnkw** [comment]
Well thank you. Indeed some of the above missing fields are rather mysterious and may not even do anything.

But other missing fields are critical. There definitely needs to be some way to specify the product images for example.

**Comment by ritikacd** [comment]
Opening again.

**Comment by johnkw** [comment]
It would be better to leave this bug report open until the bug is fixed.

**Comment by johnkw** [comment]
Oh wow I just saw this:
https://developer-docs.amazon.com/sp-api/changelog/update-deprecation-of-feeds-api-support-for-xml-and-flat-file-listings-feeds-date-changed-to-june-30-2025

Great news it's pushed back a little, but the new API is still not available with any documented migration path to use it, so that deprecation date is still rather aggressive.

**Comment by ritikacd** [comment]
Hi @johnkw 

We are working on adding missing mappings to the mapping files. It will be available soon. 

Have you checked out [Listings sample application](https://github.com/amzn/selling-partner-api-samples/tree/main/use-cases/listings)? It's an end-to-end UI application to create and manage listings workflows. You can also use the app to just generate input file for JSON_LISTINGS_FEED for the existing listings created using legacy feed types. Go to the **Update Listings** tab, enter the legacy SKU, click CONVERT TO JSON LISTINGS FEED 

![Image](https://github.com/user-attachments/assets/e62656bb-ec9e-4162-a2b6-0ba37f72f1dc)

Refer to the [Sample Video Tutorial](Sample Video Tutorial) to configure the application. Here's another video on using Listings Sample Solution for [Update Listing Flow](https://www.youtube.com/watch?v=z81k84pGpMQ). 

Let us know if this helps!

Thank You,
Ritika
Selling Partner Developer Services

**Comment by johnkw** [comment]
ETA on a fix to the docs here?

I have not looked at that web app, but does it contain mappings for all the fields that currently lack a documenting mapping?

**Comment by ritikacd** [comment]
Hi @johnkw 

We have been actively working on improving the mapping files. Please expect an update in upcoming weeks on the missing attributes. 

Thanks,
Ritika
Selling Partner API Developer Services

**Comment by johnkw** [comment]
We just got an email announcing "Important: XML and Flat File Listing Feeds Deprecation Date Updated" but it still shows the same "June 30, 2025" date as before. Is that still the date?  We're less than 2 months away, and the new API still isn't documented and whether it works or not is totally unknown.  That's not much time to prepare for a new API.  Can the date be pushed back so we have at least 3-6 months to prepare?

**Comment by ritikacd** [comment]
Hi @johnkw 

This is to notify you that the missing mappings reported have been added to the latest [Product Flat File Mappings](https://m.media-amazon.com/images/G/01/VSSC/FlatFile-To-JSON-Mapping.xlsx) file. Please review.  

Please note that the following mappings are not implemented in the JSON based Listings APIs, hence not present in the mapping file: 
back_label_image_url
front_bottle_image_url
item_dimensions_unit_of_measure
net_content_volume1
net_content_volume2
net_content_volume_unit_of_measure1
net_content_volume_unit_of_measure2
package_contains_identifier
package_contains_quantity
package_dimensions_unit_of_measure
pricing_action
progressive_discount_lower_bound1
progressive_discount_lower_bound2
progressive_discount_lower_bound3
progressive_discount_type
progressive_discount_value1
progressive_discount_value2
progressive_discount_value3

Please let us know if you are actively using those mappings. 

Thanks,
Ritika
Selling Partner API Developer Services

**Comment by ritikacd** [comment]
Hi @johnkw 

The mappings in these files are limited to product-related attributes (that is, they do not include inventory or pricing) and do not include the product type or marketplace-specific relevancy of each. The [Product Type Definitions API](https://developer-docs.amazon.com/sp-api/reference/product-type-definitions-v2020-09-01) provides JSON Schemas that represent the applicable attributes and constraints per product type and marketplace. 

Using the mapping files in conjunction with the JSON Schemas lets you update your data mappings, identify which mappings are relevant for each product type and marketplace, identify data elements no longer relevant, and identify new relevant data elements. 

Have you incorporated the Product Type Definitions API into your listings workflow to retrieve the missing attribute mappings?

Thanks,
Ritika Chaudhary
Selling Partner Developer Services

**Comment by hsbrendan** [comment]
I would like to back up what @johnkw is saying. The JSON schema is not complete and this deprecation has not been thought through yet. 

Just one example: althought the "ingredients" columns have mappings (for example, ingredients1 maps to /attributes/ingredients/0/value), errors are returned on actually using the field:

{
    "header": {
        "sellerId": "...",
        "version": "2.0",
        "feedId": "..."
    },
    "issues": [
        {
            "messageId": 1,
            "code": "90000900",
            "severity": "ERROR",
            "message": "The attributes are invalid.",
            "attributeName": "ingredients"
        },
        {
            "messageId": 2,
            "code": "90000900",
            "severity": "ERROR",
            "message": "The attributes are invalid.",
            "attributeName": "ingredients"
        }
    ],
    "summary": {
        "errors": 2,
        "warnings": 0,
        "messagesProcessed": 2,
        "messagesAccepted": 0,
        "messagesInvalid": 2
    }
}

I reached out to json-migration-support-apiis@amazon.com and have not had a reply yet.

kind regards,
Brendan

**Comment by johnkw** [comment]
> The mappings in these files are limited to product-related attributes (that is, they do not include inventory or pricing) and do not include the product type or marketplace-specific relevancy of each. The [Product Type Definitions API](https://developer-docs.amazon.com/sp-api/reference/product-type-definitions-v2020-09-01) provides JSON Schemas that represent the applicable attributes and constraints per product type and marketplace.

Thanks, but your reply makes no sense to me. For example `potassium_per_serving` is clearly one of the "product-related attributes".

This new API is not ready for general use, and the deprecation of the existing working API should be pushed out indefinitely until the new API is ready. Please advise @ritikacd ? Once the new API is ready for use a timeline should be set for deprecating the existing working API.

**Comment by Prasanjeet-Keshri-Cleo** [comment]
how to find the JSON_LISTINGS_FEED json schema that incorporates the model for all the deprecating feed types
POST_PRODUCT_DATA
POST_INVENTORY_AVAILABILITY_DATA
POST_PRODUCT_OVERRIDES_DATA
POST_PRODUCT_PRICING_DATA
POST_PRODUCT_IMAGE_DATA
POST_PRODUCT_RELATIONSHIP_DATA
POST_FLAT_FILE_INVLOADER_DATA
POST_FLAT_FILE_BOOKLOADER_DATA
POST_FLAT_FILE_CONVERGENCE_LISTINGS_DATA
POST_FLAT_FILE_LISTINGS_DATA
POST_FLAT_FILE_PRICEANDQUANTITYONLY_UPDATE_DATA
POST_UIEE_BOOKLOADER_DATA

since the JSON_LISTINGS_FEED api is replacing and in a way the json equivalent of the above xml deprecated feed types, i could really use the json schema/model for JSON_LISTINGS_FEED.

Thanks.

**Comment by johnkw** [comment]
This situation is bizarre. There's still no alternative available and the "removal" date is 3 weeks away. @ritikacd ?

**Comment by nightcoding2021** [comment]
> This situation is bizarre. There's still no alternative available and the "removal" date is 3 weeks away. [@ritikacd](https://github.com/ritikacd) ?

Personally, given the almost certain massive GMV damage/decrease, I'd like to think it's a "certainty" they will now extend the June 30 deadline.  In reality, on current plans it's still happening...

**Comment by ritikacd** [comment]
> how to find the JSON_LISTINGS_FEED json schema that incorporates the model for all the deprecating feed types

Hi @Prasanjeet-Keshri-Cleo 

Refer to the [Listings migration guide](https://developer-docs.amazon.com/sp-api/docs/listing-workflow-migration-tutorial) for JSON_LISTINGS_FEED schema examples. This guide shows how the [Listings Items API](https://developer-docs.amazon.com/sp-api/reference/listings-items-v2021-08-01) and JSON_LISTINGS_FEED relate to legacy XML feed types, outlining API endpoints and payload examples before and after migration. 

Hope this helps! 

Thanks,
Ritika

**Comment by ritikacd** [comment]
> I would like to back up what [@johnkw](https://github.com/johnkw) is saying. The JSON schema is not complete and this deprecation has not been thought through yet.
> 
> Just one example: although the "ingredients" columns have mappings (for example, ingredients1 maps to /attributes/ingredients/0/value), errors are returned on actually using the field:
> 
> { "header": { "sellerId": "...", "version": "2.0", "feedId": "..." }, "issues": [ { "messageId": 1, "code": "90000900", "severity": "ERROR", "message": "The attributes are invalid.", "attributeName": "ingredients" }, { "messageId": 2, "code": "90000900", "severity": "ERROR", "message": "The attributes are invalid.", "attributeName": "ingredients" } ], "summary": { "errors": 2, "warnings": 0, "messagesProcessed": 2, "messagesAccepted": 0, "messagesInvalid": 2 } }
> 
> I reached out to [json-migration-support-apiis@amazon.com](mailto:json-migration-support-apiis@amazon.com) and have not had a reply yet.
> 
> kind regards, Brendan

Hi @hsbrendan 

Please submit a troubleshooting ticket with the developer support team using the guidelines listed [here](https://developer.amazonservices.com/support). 

Thank you,
Ritika

**Comment by ritikacd** [comment]
> > The mappings in these files are limited to product-related attributes (that is, they do not include inventory or pricing) and do not include the product type or marketplace-specific relevancy of each. The [Product Type Definitions API](https://developer-docs.amazon.com/sp-api/reference/product-type-definitions-v2020-09-01) provides JSON Schemas that represent the applicable attributes and constraints per product type and marketplace.
> 
> Thanks, but your reply makes no sense to me. For example `potassium_per_serving` is clearly one of the "product-related attributes".
> 
> This new API is not ready for general use, and the deprecation of the existing working API should be pushed out indefinitely until the new API is ready. Please advise [@ritikacd](https://github.com/ritikacd) ? Once the new API is ready for use a timeline should be set for deprecating the existing working API.

Hi @johnkw 

We recommend using the [Product Type Definitions API](https://developer-docs.amazon.com/sp-api/docs/product-type-api-use-case-guide) if you find attributes missing in the mapping files. This API provides the most up-to-date data requirements and attribute definitions by product type and marketplace ID.

Have you already called the API to verify the expected attributes? 


Thanks,
Ritika
Selling Partner Developer Services

**Comment by johnkw** [comment]
@ritikacd once again I don't understand this comment at all. The new API is not yet ready for use, but the current working API is now scheduled for removal in 2 weeks. This makes no sense. It requires IMMEDIATE attention to get  the current working API removed from the deprecation/removal schedule.

As I listed above in https://github.com/amzn/selling-partner-api-models/issues/4318#issuecomment-2902140899 there are currently dozens of fields which have no mapping to the new API.  Until those (at a minimum) are fixed, the new API is not ready for general use.

**Comment by elburro1887** [comment]
@ritikacd I would also advise to push back the deprecation date. There is no clear path to migration, and, I believe that a majority of your sellers will not be able to migrate so fast. 

How can there be dozens of fields missing, 2 weeks before the deprecation date?

This new JSON based system is too complicated, not well thought out and too broken to have such an aggressive migration date.

**Comment by hsbrendan** [comment]
@ritikacd I am in contact with the developer support team. They replied last week; however, their solution did not work at all.

Nobody at the amazon team seems to care that on 30 Jun, many larger sellers will be completely unable to maintain their feeds.

The JSON schema is not ready.

**Comment by simoncc** [comment]
There is now only a week to go until the new feed requirement is enforced. We have our own issues understanding the correct format to submit business pricing and Discount pricing, with conflicting information on these in the documentation.

Truly hoping an extension will be granted, but seen nothing to suggest it will be.

**Comment by johnkw** [comment]
In one of the prior bug reports (unfortunately not sure where after all the github insanity of moving repos for the API and a lot of bugs getting inappropriately closed) Amazon stated that the sellercentral web GUI upload of CSV files would still be allowed even after the API upload of CSV files became disallowed.  So that is apparently the only fallback option currently.  It will certainly be more error-prone and problematic than direct API use.

**Comment by nightcoding2021** [comment]
Amazon has not announced that TAB file uploads will be blocked via Seller Central, and probably they wouldn't ever be able to do that - because many sellers would use spreadsheets.    Due to JSON limits and the ridiclously slow processing, in order to even make price updates affecting entire inventory I would have to use manual Seller Central uploads.   Going from where we are now, to that, is definitely not an improvement but a severe business downgrade being done voluntarily by Amazon.

However, from experience, if a business/corporation wants to make itself worse off - they are able/entitled to do so.  Just look at the downward spiral disaster that eBay got itself locked into since GTC.

**Comment by ritikacd** [comment]
> There is now only a week to go until the new feed requirement is enforced. We have our own issues understanding the correct format to submit business pricing and Discount pricing, with conflicting information on these in the documentation.

Hi @simoncc 
Can you please point out what you observed is conflicting in the documentation? 

Thanks,
Ritika

**Comment by ritikacd** [comment]
> [@ritikacd](https://github.com/ritikacd) I am in contact with the developer support team. They replied last week; however, their solution did not work at all.
> 
> Nobody at the amazon team seems to care that on 30 Jun, many larger sellers will be completely unable to maintain their feeds.
> 
> The JSON schema is not ready.

Can you please share the case id?

**Comment by simoncc** [comment]
> > There is now only a week to go until the new feed requirement is enforced. We have our own issues understanding the correct format to submit business pricing and Discount pricing, with conflicting information on these in the documentation.
> 
> Hi [@simoncc](https://github.com/simoncc) Can you please point out what you observed is conflicting in the documentation?
> 
> Thanks, Ritika

For instance on this page pricing feed mapping

https://developer-docs.amazon.com/sp-api/docs/mapping-product-attributes#pricing-feed-mappings

Discounted price schedule start date. Shows 
/attributes/purchasable_offer/0/discounted_price/0/schedule/0/start_at

but is it not

/attributes/purchasable_offer/0/our_price/0/schedule/0/schedule_time_period.start_at

I am confused as to which is the correct syntax compatible with JSON_LISTINGS_FEED I am basically looking for actual examples of JSON-formatted files for updating Price, Quantity, Discounted Price, and Business Price with Quantity discounts.

**Comment by nightcoding2021** [comment]
Hopefully you can get that resolved with a staffer on the thread.  However, having been through this myself, it's obvious this is nowhere near a point where TAB/XML files should be switched off.  With no announcement to the contrary, they are rushing ahead anyway and damage will be caused.

**Comment by johnkw** [comment]
@ritikacd what's the status on getting the deprecation indefinitely postponed, until such time as there's a documented path to using the new API (which may or may not even work at all, since no one can even beta test it yet without a documented migration path)?

**Comment by nightcoding2021** [comment]
This announcement is going out currently via email - now extended to July 31.  At least that gives breathing space to everyone worrying about June 30.  Anyone starting this migration late has no chance.  I started about 6 months ago and it was a massive burden/drain on resources for months.

Not sure if they posted in announcements yet.

![Image](https://github.com/user-attachments/assets/7aec097f-9bda-4dac-b994-e9edc9edb66e)

**Comment by johnkw** [comment]
Ah well that's better than nothing. July 31, 2025 provides um ~5 weeks or so. But there's still no documented conversion path.

At a minimum there needs to be a FlatFile-To-JSON-Mapping.xlsx file with all of the fields defined in it.

**Comment by ritikacd** [comment]
> Here are the list of required fields which don't have a mapping:
> 
> ```
> added_sugars_per_serving
> are_batteries_included
> batteries_required
> calcium_per_serving
> calcium_per_serving_string
> calcium_per_serving_unit_of_measure
> container_type
> contains_liquid_contents
> country_string
> dietary_fiber_per_serving
> dietary_fiber_per_serving_string
> directions
> each_unit_count
> energy_content_from_trans_fat_per_serving
> energy_content_per_serving
> energy_content_per_serving_string
> energy_content_unit_of_measure
> fat_unit_of_measure
> fc_shelf_life
> flavor_name
> generic_keywords
> iron_per_serving
> iron_per_serving_string
> iron_per_serving_unit_of_measure
> is_expiration_dated_product
> is_heat_sensitive
> item_form
> item_package_quantity
> item_type
> liquid_packaging_type
> list_price
> manufacturer
> number_of_items
> package_dimensions_unit_of_measure
> package_height
> package_height_unit_of_measure
> package_length
> package_length_unit_of_measure
> package_type_name
> package_weight
> package_weight_unit_of_measure
> package_width
> package_width_unit_of_measure
> parent_child
> pattern_name
> potassium_per_serving
> potassium_per_serving_string
> potassium_per_serving_unit_of_measure
> product_expiration_type
> product_site_launch_date
> protein_per_serving
> protein_per_serving_string
> protein_per_serving_unit_of_measure
> region_of_origin
> saturated_fat_per_serving_string
> serving_size
> serving_size_description
> serving_size_unit_of_measure
> size_name
> sodium_per_serving
> sodium_per_serving_string
> sodium_per_serving_unit_of_measure
> sugar_alcohols_per_serving
> sugar_alcohols_per_serving_string
> sugars_per_serving
> sugars_per_serving_string
> temperature_rating
> total_carbohydrate_per_serving
> total_carbohydrate_per_serving_string
> total_fat_per_serving
> total_fat_per_serving_string
> trans_fat_per_serving
> unit_count
> unit_count_type
> variation_theme
> ```
> 
> Can we get an extension on the deprecation date for the current working method, since the new method is not yet ready? At a minimum it seems to need to support the above missing fields.

Hi @johnkw 

We were able to locate the mappings for all of the above attributes in the current product flat file. Some may be listed under different names. For your reference, I've attached the file here with above mappings. I hope this helps!

[FlatFile-To-JSON-Mapping.xlsx](https://github.com/user-attachments/files/21029482/FlatFile-To-JSON-Mapping.xlsx)

As you update your workflows, we highly recommend reviewing the provided use case guide to follow best practices:

- [Building Listings Management Workflows Guide](https://developer-docs.amazon.com/sp-api/docs/building-listings-management-workflows-guide)
- [Listings Management Workflow Migration](https://developer-docs.amazon.com/sp-api/docs/listings-management-workflow-migration)


Thanks,
Ritika

**Comment by johnkw** [comment]
Thanks, however the new API is not yet ready as the following are still missing from the above xlsx file:

```
allergen_information1
allergen_information2
allergen_information3
allergen_information4
allergen_information5
brand_name
bullet_point1
bullet_point2
bullet_point3
bullet_point4
bullet_point5
condition_type
country_of_origin
cpsia_cautionary_statement1
diet_type1
diet_type2
external_product_id
external_product_id_type
feed_product_type
fulfillment_center_id
ingredients1
item_name
item_sku
item_weight
item_weight_unit_of_measure
main_image_url
material_type_free1
material_type_free2
material_type_free3
material_type_free4
material_type_free5
max_order_quantity
merchant_shipping_group_name
other_image_url1
other_image_url2
other_image_url3
other_image_url4
other_image_url5
parent_sku
product_description
relationship_type
serving_recommendation
specialty1
specialty2
specialty3
specialty4
specialty5
standard_price
supplier_declared_dg_hz_regulation1
target_audience_keywords1
target_audience_keywords2
target_audience_keywords3
target_audience_keywords4
target_audience_keywords5
thesaurus_subject_keywords1
thesaurus_subject_keywords2
update_delete
```

**Comment by ritikacd** [comment]
> Thanks, however the new API is not yet ready as the following are still missing from the above xlsx file:
> 
> ```
> allergen_information1
> allergen_information2
> allergen_information3
> allergen_information4
> allergen_information5
> brand_name
> bullet_point1
> bullet_point2
> bullet_point3
> bullet_point4
> bullet_point5
> condition_type
> country_of_origin
> cpsia_cautionary_statement1
> diet_type1
> diet_type2
> external_product_id
> external_product_id_type
> feed_product_type
> fulfillment_center_id
> ingredients1
> item_name
> item_sku
> item_weight
> item_weight_unit_of_measure
> main_image_url
> material_type_free1
> material_type_free2
> material_type_free3
> material_type_free4
> material_type_free5
> max_order_quantity
> merchant_shipping_group_name
> other_image_url1
> other_image_url2
> other_image_url3
> other_image_url4
> other_image_url5
> parent_sku
> product_description
> relationship_type
> serving_recommendation
> specialty1
> specialty2
> specialty3
> specialty4
> specialty5
> standard_price
> supplier_declared_dg_hz_regulation1
> target_audience_keywords1
> target_audience_keywords2
> target_audience_keywords3
> target_audience_keywords4
> target_audience_keywords5
> thesaurus_subject_keywords1
> thesaurus_subject_keywords2
> update_delete
> ```

Hi @johnkw 

The above mapping file was provided for the missing attributes reported in the earlier comment. To see the complete list of attribute, refer to the [Product Flat File Mappings](https://m.media-amazon.com/images/G/01/VSSC/FlatFile-To-JSON-Mapping.xlsx) provided in the listings migration guide. 

Thanks, 
Ritika

**Comment by johnkw** [comment]
@ritikacd Are you implying that users need to somehow combine multiple FlatFile-To-JSON-Mapping.xlsx files from different locations and use them together in order to get a functional mapping to use the new API?

**Comment by ritikacd** [comment]
> [@ritikacd](https://github.com/ritikacd) Are you implying that users need to somehow combine multiple FlatFile-To-JSON-Mapping.xlsx files from different locations and use them together in order to get a functional mapping to use the new API?

Hi @johnkw 

No, we recommend relying on the files provided in the [migration guide](https://developer-docs.amazon.com/sp-api/docs/mapping-product-attributes). As noted [here](https://github.com/amzn/selling-partner-api-models/issues/4318#issuecomment-3030575208), some mappings may appear under different names. To help with that, we’ve already pulled those mappings from the existing file for your convenience.

Could you help me understand how you're using these mapping files? Are you relying solely on them to migrate your data flows? 

Thanks,
Ritika

**Comment by johnkw** [comment]
I can't make any sense of your reply. Where is the documentation on how to map, specifically for example, just `other_image_url1` which is completely missing from the FlatFile-To-JSON-Mapping.xlsx file?

**Comment by ritikacd** [comment]
> I can't make any sense of your reply. Where is the documentation on how to map, specifically for example, just `other_image_url1` which is completely missing from the FlatFile-To-JSON-Mapping.xlsx file?

@johnkw 
I was able to find `other_image_url1` in the [Product Flat File Mappings](https://m.media-amazon.com/images/G/01/VSSC/FlatFile-To-JSON-Mapping.xlsx) available on the [Mapping Product Attribute page](https://developer-docs.amazon.com/sp-api/docs/mapping-product-attributes). 

<img width="2836" height="1560" alt="Image" src="https://github.com/user-attachments/assets/c79883fc-18ff-4f65-b504-6374a248307e" />

<img width="1065" height="676" alt="Image" src="https://github.com/user-attachments/assets/1da816ac-dc9e-4603-80f8-e63e8e587f71" />

Thanks,
Ritika

**Comment by johnkw** [comment]
Ok if I use https://m.media-amazon.com/images/G/01/VSSC/FlatFile-To-JSON-Mapping.xlsx then this is the list of missing fields:

```added_sugars_per_serving
are_batteries_included
batteries_required
calcium_per_serving
calcium_per_serving_string
calcium_per_serving_unit_of_measure
container_type
country_string
dietary_fiber_per_serving
dietary_fiber_per_serving_string
directions
each_unit_count
energy_content_from_trans_fat_per_serving
energy_content_per_serving
energy_content_per_serving_string
energy_content_unit_of_measure
fat_unit_of_measure
fc_shelf_life
flavor_name
generic_keywords
iron_per_serving
iron_per_serving_string
iron_per_serving_unit_of_measure
is_expiration_dated_product
is_heat_sensitive
item_form
item_package_quantity
item_type
list_price
manufacturer
number_of_items
package_height
package_height_unit_of_measure
package_length
package_length_unit_of_measure
package_type_name
package_weight
package_weight_unit_of_measure
package_width
package_width_unit_of_measure
parent_child
pattern_name
potassium_per_serving
potassium_per_serving_string
potassium_per_serving_unit_of_measure
product_expiration_type
product_site_launch_date
protein_per_serving
protein_per_serving_string
protein_per_serving_unit_of_measure
region_of_origin
saturated_fat_per_serving_string
serving_recommendation
serving_size
serving_size_description
serving_size_unit_of_measure
sodium_per_serving
sodium_per_serving_string
sodium_per_serving_unit_of_measure
sugar_alcohols_per_serving
sugar_alcohols_per_serving_string
sugars_per_serving
sugars_per_serving_string
temperature_rating
total_carbohydrate_per_serving
total_carbohydrate_per_serving_string
total_fat_per_serving
total_fat_per_serving_string
trans_fat_per_serving
unit_count
unit_count_type
variation_theme
```

What's the status on extending the deprecation date until such time as the new API is ready?

**Comment by johnkw** [comment]
Any update on when the documentation will be fixed to possibly migrate? The deprecation date is now tomorrow, and there's still no way to migrate to the new feed. @ritikacd @puppsupr

**Comment by ritikacd** [comment]
> Ok if I use https://m.media-amazon.com/images/G/01/VSSC/FlatFile-To-JSON-Mapping.xlsx then this is the list of missing fields:
> 
> ```
> are_batteries_included
> batteries_required
> calcium_per_serving
> calcium_per_serving_string
> calcium_per_serving_unit_of_measure
> container_type
> country_string
> dietary_fiber_per_serving
> dietary_fiber_per_serving_string
> directions
> each_unit_count
> energy_content_from_trans_fat_per_serving
> energy_content_per_serving
> energy_content_per_serving_string
> energy_content_unit_of_measure
> fat_unit_of_measure
> fc_shelf_life
> flavor_name
> generic_keywords
> iron_per_serving
> iron_per_serving_string
> iron_per_serving_unit_of_measure
> is_expiration_dated_product
> is_heat_sensitive
> item_form
> item_package_quantity
> item_type
> list_price
> manufacturer
> number_of_items
> package_height
> package_height_unit_of_measure
> package_length
> package_length_unit_of_measure
> package_type_name
> package_weight
> package_weight_unit_of_measure
> package_width
> package_width_unit_of_measure
> parent_child
> pattern_name
> potassium_per_serving
> potassium_per_serving_string
> potassium_per_serving_unit_of_measure
> product_expiration_type
> product_site_launch_date
> protein_per_serving
> protein_per_serving_string
> protein_per_serving_unit_of_measure
> region_of_origin
> saturated_fat_per_serving_string
> serving_recommendation
> serving_size
> serving_size_description
> serving_size_unit_of_measure
> sodium_per_serving
> sodium_per_serving_string
> sodium_per_serving_unit_of_measure
> sugar_alcohols_per_serving
> sugar_alcohols_per_serving_string
> sugars_per_serving
> sugars_per_serving_string
> temperature_rating
> total_carbohydrate_per_serving
> total_carbohydrate_per_serving_string
> total_fat_per_serving
> total_fat_per_serving_string
> trans_fat_per_serving
> unit_count
> unit_count_type
> variation_theme
> ```
> 
> What's the status on extending the deprecation date until such time as the new API is ready?

@johnkw Here's the merged flat file to JSON mapping list:  

[FlatFile-To-JSON-Mapping-Final.xlsx](https://github.com/user-attachments/files/21515835/FlatFile-To-JSON-Mapping-Final.xlsx)

**Comment by johnkw** [comment]
Thanks @ritikacd that's a huge improvement. I now only see one missing column from the FoodAndBeverages category: `serving_recommendation`. Do you know what `serving_recommendation` got moved to in the new API?

**Comment by ritikacd** [comment]
> Thanks [@ritikacd](https://github.com/ritikacd) that's a huge improvement. I now only see one missing column from the FoodAndBeverages category: `serving_recommendation`. Do you know what `serving_recommendation` got moved to in the new API?

Yes, it should be present in the JSON APIs. Here's the mapping: `/attributes/serving_recommendation/0/value`.

**Comment by johnkw** [comment]
Ah ok, thanks, so if I manually add that entry to the FlatFile-To-JSON-Mapping-Final file, then a mapping now exists for every field, at least for the FoodAndBeverages category. The documentation page needs to be fixed to include all those additional fields you added.

However, now looking at the next steps here, there appears to be no documented migration path regarding product types.

The current working feed API has a column called `feed_product_type` and also a column called `item_type`, however neither of those appear to be the new https://developer-docs.amazon.com/sp-api/reference/putlistingsitem API's `productType`. How can we migrate/map those over?

On the https://developer-docs.amazon.com/sp-api/reference/putlistingsitem page the documentation needs to be fixed to say where `productType` comes from, and how it can be mapped over from the current feed files from Seller Central.

(By the way, it is possible to find the product type for a SKU that already exists, using `searchListingsItems`. But for the creation of a new SKU, starting with the Seller Central self-documenting template, there appears to be no way to translate from one to the other.)

**Comment by johnkw** [comment]
There appear to be some errors in the mapping file. I made these changes which seem to fix some breakages.

```
--- FlatFile-To-JSON-Mapping.orig.csv   2025-07-31 19:23:15.984713036 +0000
+++ FlatFile-To-JSON-Mapping.csv        2025-07-31 20:27:03.480811122 +0000
@@ -2286 +2286 @@
-monounsaturated_fat_unit_of_measure            /attributes/nutritional_info/0/fat/unit
+monounsaturated_fat_unit_of_measure            /attributes/nutritional_info/0/fat/0/unit
@@ -2920 +2920 @@
-saturated_fat_unit_of_measure          /attributes/nutritional_info/0/fat/unit
+saturated_fat_unit_of_measure          /attributes/nutritional_info/0/fat/0/unit
@@ -9179 +9178,0 @@
-
@@ -9197 +9196 @@
-fat_unit_of_measure            /attributes/nutritional_info/0/fat/unit
+fat_unit_of_measure            /attributes/nutritional_info/0/fat/0/unit
@@ -9233 +9232 @@
-saturated_fat_per_serving_string               /attributes/nutritional_info/0/fat/unit
+saturated_fat_per_serving_string               /attributes/nutritional_info/0/fat/0/unit
```

(This new API sure is bizarrely clunky and needlessly complicated. So many pointless nested layers of dictionaries and lists.)

The above issues are similar to those reported in bug #4264. It's possible there are more of these, but these are just the ones I noticed.

**Comment by jlevers** [comment]
I realized a while ago while making my [feed transformer tool](https://tools.highsidelabs.co/) that it is technically possible to exhaustively error check the field mappings, but I haven't gotten around to actually writing the code to do that...I don't think it should take more than a few days, if you guys (Amazon devs) would do that it would be really really helpful

**Comment by johnkw** [comment]
It's weird that there would be errors in that file at all. It seems like something that would be automatically exported from whatever internal mapping system that's being used for the normal XLSX feed file uploads on Seller Central. It almost seems like it was somehow hand-crafted by someone though.

**Comment by johnkw** [comment]
I just found another missing column: `saturated_fat_per_serving` -- what is that one @ritikacd ?

**Comment by johnkw** [comment]
Oh I think maybe it's hidden inside a bizarrely complicated list of dictionaries in `/attributes/nutritional_info/0/vitamins_and_minerals`. Each one seems to get a separate row with a name that's generated mysteriously from some other mapping. I just realized all of those are being incorrectly collided onto the same path.

Looking at an example `getListingsItem` call, it appears the format of this section is a list of dictionaries where the identifier is hidden in a field called "nutrient" instead of just being the dictionary key for that section as one might expect. What a bizarrely complicated mess.
```
    "vitamins_and_minerals": [
     {
      "nutrient": "calcium",
      "string_value": "110",
      "unit": "milligrams",
      "value": "110.00"
     },
     {
      "nutrient": "iron",
      "string_value": "2.7",
      "unit": "milligrams",
      "value": "2.7"
     },
     {
      "nutrient": "potassium",
      "string_value": "880",
      "unit": "milligrams",
      "value": "880.00"
     },
     {
      "nutrient": "sodium",
      "string_value": "0.95",
      "unit": "grams",
      "value": "0.95"
     }
    ],
```
It's not obvious how this could even be described in the form of a JSON path. (Assuming they wanted to do all of these pointlessly complicated nested structures, what a huge mistake they made not to just to make it be a dictionary like `"vitamins_and_minerals": { "calcium": { ... } }` which could at least then be expressed in the form of the path.)

But can we somehow assume that whenever anything in the `/attributes/nutritional_info/0/vitamins_and_minerals` is used, that you can just magically chop off the `_per_serving` part of `calcium_per_serving` and use that as the `nutrient` key?

It's worth nothing that it was only a couple days ago the a migration path was even documented at all, and the deprecation date has already passed. The deprecation date for the current feeds should be pushed out at least a few months to give time to look at all this now that there's some documentation on migration.

**Comment by nightcoding2021** [comment]
They announced this a few hours ago, despite claiming yesterday they'd already turned off TAB uploads.

I've switched to JSON (but with massively restricted updates due to the lower processing capacity of JSON feeds compared to TAB previously).

Some sellers might be exiting Amazon selling altogether as a result of this change.  I did manage to migrate but not without massive effort.  If my situation was different, that might have been impossible and I'd have thrown in the towel when TAB files were ended.

Sellers can still upload TAB files via Seller Central.  What a strange idea turning them off via the API, if still available via Seller Central web upload!

<img width="593" height="582" alt="Image" src="https://github.com/user-attachments/assets/29ef1227-8913-439f-895d-98ebcfec29a3" />

**Comment by johnkw** [comment]
@ritikacd @puppsupr Any word on those migration issues?

**Comment by johnkw** [comment]
Please reopen @ritikacd @puppsupr ? This is still an active critical issue so long as `POST_FLAT_FILE_LISTINGS_DATA` is deprecated.

**Comment by johnkw** [comment]
See also bug #4456.

**Comment by ritikacd** [comment]
> See also bug [#4456](https://github.com/amzn/selling-partner-api-models/issues/4456).

Hi @johnkw 
The reported bug was due to incorrect and duplicated records in the XPath-to-JSON pointer mappings within the mapping file. Please note that these mapping files were last updated on May 27. Since SP-API no longer supports submissions through legacy feed types, we won’t be making further updates to the existing migration documents.

If you have any questions regarding the JSON-based Listings APIs, please open a new issue and reach out to the SP-API [developer support services](https://developer.amazonservices.com/support-guidelines) team. 

Thank you,
Ritika

**Comment by johnkw** [comment]
The lack of a migration path is still relevant right now. It has nothing to do with whether `POST_FLAT_FILE_LISTINGS_DATA` is supported in the API or not.

All of the documentation for the product listings and what the meaning of the attributes are is contained within the Seller Central feed files.  There is no other working documentation for that.  Therefore a migration path from the simple feed column names to the complicated JSON hierarchy is the only way to get documentation on the complicated JSON hierarchy.

Feel free to point out some other way?

Developer support is unable to provide support on the missing documentation.  Please reopen this issue?

**Comment by ritikacd** [comment]
> The lack of a migration path is still relevant right now. It has nothing to do with whether `POST_FLAT_FILE_LISTINGS_DATA` is supported in the API or not.
> 
> All of the documentation for the product listings and what the meaning of the attributes are is contained within the Seller Central feed files. There is no other working documentation for that. Therefore a migration path from the simple feed column names to the complicated JSON hierarchy is the only way to get documentation on the complicated JSON hierarchy.
> 
> Feel free to point out some other way?
> 
> Developer support is unable to provide support on the missing documentation. Please reopen this issue?

Attributes, data requirements and their meanings can now be retrieved through the [Product Type Definitions API](https://developer-docs.amazon.com/sp-api/reference/product-type-definitions-v2020-09-01).

**Comment by johnkw** [comment]
> Attributes, data requirements and their meanings can now be retrieved through the [Product Type Definitions API](https://developer-docs.amazon.com/sp-api/reference/product-type-definitions-v2020-09-01).

That would be nice if it were true, but it's not true.  There is no associated documentation on the meaning of anything in there.  And if you try to ask Seller Support for information, Seller Support will only provide support on the feed files, which is the only method of bulk updates which currently works correctly.

Go to https://www.amazon.com/dp/B002OTJO9I and click "Nutrition facts" and see where it says "2% Calcium"?  Now try to find for example where in the complicated JSON hierarchy it is that you set the 2% Calcium number for an ASIN.  It's not in the new Listing SPAPI documentation.  You won't be able to find it.  And it must be there, before new Listing SPAPI can be production-ready.

Please reopen this issue?

**Comment by ritikacd** [comment]
> > Attributes, data requirements and their meanings can now be retrieved through the [Product Type Definitions API](https://developer-docs.amazon.com/sp-api/reference/product-type-definitions-v2020-09-01).
> 
> That would be nice if it were true, but it's not true. There is no associated documentation on the meaning of anything in there. And if you try to ask Seller Support for information, Seller Support will only provide support on the feed files, which is the only method of bulk updates which currently works correctly.
> 
> Go to https://www.amazon.com/dp/B002OTJO9I and click "Nutrition facts" and see where it says "2% Calcium"? Now try to find for example where in the complicated JSON hierarchy it is that you set the 2% Calcium number for an ASIN. It's not in the new Listing SPAPI documentation. You won't be able to find it. And it must be there, before new Listing SPAPI can be production-ready.
> 
> Please reopen this issue?

Product Type for that ASIN is PACKAGED_SOUP_AND_STEW. By calling the product type definitions APIs for that product type, you can find the `vitamins_and_minerals` attribute, which allows providing the nutrients and its value. 

Another way to see attribute names for an existing ASIN would be if you call getCatalogItems API for that ASIN with includedData= attributes, summaries, productTypes, etc.. The response returns attribute information. See example below. 

<img width="1022" height="646" alt="Image" src="https://github.com/user-attachments/assets/9089b85f-5f85-42c7-9912-8debe7c4c726" />

I would also recommend checking out our [Postman collection](https://www.postman.com/amazon-selling-partner-api/sp-api/collection/g9ywc0f/listings-apis-workflows) for the Listings workflow, which will help you with the step-by-step guidance for managing your listings.

I hope this helps! 

Thank you,
Ritika

**Comment by johnkw** [comment]
Thanks for the reply.

So, the https://sellingpartnerapi-na.amazon.com/definitions/2020-09-01/productTypes/ schema contains a vague undocumented "vitamins_and_minerals" attribute in the complicated JSON hierarchy, but there's no documentation on what it is and what the valid values are.  In Seller Central the feeds are all documented with that information.

How can you look at the /productTypes/ schema and possibly know precisely what the allow values and units are for that?

Until there's a schema with documentation on what goes into it, and some way to contact Seller Support about issues in it, there is no migration path to use `JSON_LISTINGS_FEED` and the new Listings API is not production-ready.

**Comment by ritikacd** [comment]
> Thanks for the reply.
> 
> So, the https://sellingpartnerapi-na.amazon.com/definitions/2020-09-01/productTypes/ schema contains a vague undocumented "vitamins_and_minerals" attribute in the complicated JSON hierarchy, but there's no documentation on what it is and what the valid values are. In Seller Central the feeds are all documented with that information.
> 
> How can you look at the /productTypes/ schema and possibly know precisely what the allow values and units are for that?
> 
> Until there's a schema with documentation on what goes into it, and some way to contact Seller Support about issues in it, there is no migration path to use `JSON_LISTINGS_FEED` and the new Listings API is not production-ready.

Hi @johnkw 

The purpose of the Product Type Definitions API is to provide that granularity to users. For example, if you observe in the below example for vitamins_and_minerals -> nutrient the acceptable values are any String or one of the enum values you see under "Nutrient". Similarly, the acceptable unit values for "Vitamin and Mineral Unit" are those under "unit".

```
{
  "vitamins_and_minerals": {
    "title": "Vitamins And Minerals",
    "description": "Provide the amount of vitamins and minerals in the product",
    "examples": [
      "610 milligrams Sodium"
    ],
    "type": "array",
    "minItems": 1,
    "minUniqueItems": 1,
    "maxUniqueItems": 30,
    "selectors": [
      "nutrient"
    ],
    "items": {
      "type": "object",
      "required": [
        "nutrient",
        "unit"
      ],
      "properties": {
        "nutrient": {
          "title": "Nutrient",
          "description": "Provide the name of the nutrient",
          "editable": true,
          "hidden": false,
          "examples": [
            "Calcium, Iron, Sodium"
          ],
          "type": "string",
          "anyOf": [
            {
              "type": "string"
            },
            {
              "type": "string",
              "enum": [
                "Biotin",
                "Calcium",
                "Chloride",
                "Cholesterol",
                "Chromium",
                "Copper",
                "Fluoride",
                "Folic Acid",
                "Iodine",
                "Iron",
                "Magnesium",
                "Manganese",
                "Molybdenum",
                "Niacin",
                "Pantothenic Acid",
                "Phosphorus",
                "Potassium",
                "Salt",
                "Selenium",
                "Sodium",
                "Thiamin",
                "Vitamin A",
                "Vitamin B12",
                "Vitamin B2",
                "Vitamin B6",
                "Vitamin C",
                "Vitamin D",
                "Vitamin E",
                "Vitamin K",
                "Zinc"
              ],
              "enumNames": [
                "Biotin",
                "Calcium",
                "Chloride",
                "Cholesterol",
                "Chromium",
                "Copper",
                "Fluoride",
                "Folic Acid",
                "Iodine",
                "Iron",
                "Magnesium",
                "Manganese",
                "Molybdenum",
                "Niacin",
                "Pantothenic Acid",
                "Phosphorus",
                "Potassium",
                "Salt",
                "Selenium",
                "Sodium",
                "Thiamin",
                "Vitamin A",
                "Vitamin B12",
                "Vitamin B2",
                "Vitamin B6",
                "Vitamin C",
                "Vitamin D",
                "Vitamin E",
                "Vitamin K",
                "Zinc"
              ]
            }
          ]
        },
        "string_value": {
          "title": "Vitamins and Minerals String Value",
          "description": "Provide the vitamin or mineral nutrient and amount that is in the product (alphanumeric)",
          "editable": true,
          "hidden": false,
          "examples": [
            "610 milligrams Sodium"
          ],
          "type": "string",
          "maxUtf8ByteLength": 5000
        },
        "unit": {
          "title": "Vitamin and Mineral Unit",
          "description": "Select the corresponding unit",
          "editable": true,
          "hidden": false,
          "examples": [
            "milligrams, grams"
          ],
          "type": "string",
          "enum": [
            "ampules",
            "food_bars",
            "bars",
            "capsules",
            "fluid_ounces",
            "grams",
            "micrograms",
            "milligrams",
            "milliliters",
            "percent_daily_value_fda",
            "pills",
            "portions",
            "scoops",
            "teaspoons"
          ],
          "enumNames": [
            "ampule(s)",
            "bar(s)",
            "Bars",
            "capsule(s)",
            "Fluid Ounces",
            "Grams",
            "Micrograms",
            "Milligrams",
            "Milliliters",
            "Percent Daily Value",
            "pill(s)",
            "Portion(s)",
            "scoop(s)",
            "teaspoon(s)"
          ]
        },
        "value": {
          "title": "Vitamins And Minerals",
          "description": "Provide the amount of the vitamin or mineral nutrient that is in the product",
          "editable": true,
          "hidden": false,
          "examples": [
            "610"
          ],
          "type": "number",
          "maxLength": 5000
        }
      },
      "additionalProperties": false
    }
  }
}
```

Also, could you please share the links in Seller Central where feeds are all documented with the attribute information? 

Thanks,
Ritika

**Comment by johnkw** [comment]
None of that post above tells you what it actually *IS*.  Is it calcium per ASIN, per 100g, per serving, what?

The simple feed file name is `calcium_per_serving_string` which is somewhat self-documenting.  But more importantly, if you have a question as far as what `calcium_per_serving_string` is supposed to be, Seller Support can answer.

Without a migration path to and from the complicated (pointlessly complicated, if I may editorialize for a moment) JSON hierarchical paths, Seller Support can't give you any information.

**Comment by johnkw** [comment]
Any word on a documentation path? The feeds still work great, but require a pointless manual step every time. This should be reopened until such time as a documentation upgrade path exists.
